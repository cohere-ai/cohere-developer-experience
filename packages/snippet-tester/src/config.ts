import path from "path";

export interface Language {
    name: string;
    exec: (file: string) => string;
    snippetRoot: (base: string) => string;
    glob: string;
}

export const config: {
    languages: Language[];
} = {
    "languages": [
        {
            "name": "node",
            "exec": (file: string) => `node ${file}`,
            "snippetRoot": (base: string) => path.join(base, "./snippets/node"),
            "glob": "**/*.js"
        },
        // {
        //     "name": "java",
        //     "exec": (file: string) => `java ${file}`,
        //     "snippetRoot": (base: string) => path.join(base, "./snippets/java"),
        //     "glob": "**/*.java"
        // },
        // {
        //     "name": "python",
        //     "exec": (file: string) => `python ${file}`,
        //     "snippetRoot": (base: string) => path.join(base, "./snippets/python"),
        //     "glob": "**/*.go"   
        // },
        // {
        //     "name": "golang",
        //     "exec": (file: string) => `go run ${file}`,
        //     "snippetRoot": (base: string) => path.join(base, "./snippets/go"),
        //     "glob": "**/*.go"
        // }
    ]
}