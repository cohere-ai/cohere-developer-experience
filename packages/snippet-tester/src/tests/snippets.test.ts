import { describe, expect, test } from '@jest/globals';
import { CohereClient } from 'cohere-ai';
import * as fs from "fs";
import { glob } from "glob";
import path from 'path';
import { config, Language } from '../config';
import { execCmd } from '../utils';

// recurse up until you find a pnpm-workspace.yaml  
const findPnpmWorkspaceDir = (curDir = __dirname) => {
    if (curDir === "/") {
        throw new Error("Could not find pnpm-workspace.yaml")
    }

    if (fs.existsSync(path.resolve(curDir, "pnpm-lock.yaml"))) {
        return curDir
    }

    return findPnpmWorkspaceDir(fs.realpathSync(path.resolve(curDir, "..")))
}

const workspaceDir = findPnpmWorkspaceDir()

const failureReasons = {
    "NotFoundError": {
        "description": "Thrown when a resource is not found. It occurs in various contexts, such as embed jobs, datasets, connectors, and finetuned models."
    },
    "BadRequestError": {
        "description": "Indicates a bad request due to invalid parameters or data. Seen in connector creation, finetuned model operations, and dataset-related actions."
    },
    "InternalServerError": {
        "description": "Represents an internal server error, suggesting an issue with server-side processing. Appears in listing events for finetuned models."
    },
    "JsonError": {
        "description": "Related to JSON parsing issues, occurs when there is a problem with JSON data, such as missing required keys."
    },
    "TypeError": {
        "description": "Thrown when there is a type-related issue, often due to unexpected keyword arguments or incorrect data types."
    },
    "FileNotFoundError": {
        "description": "Indicates that a file is not found, as seen in the dataset-post.py test."
    },
    "IllegalStateException": {
        "description": "Related to the state of an object, occurs due to a closed resource."
    },
    "ReferenceError": {
        "description": "Thrown when a variable or object is accessed before it is defined or initialized."
    },
    "SocketTimeoutException": {
        "description": "Indicates a network timeout, a network operation took longer than the specified timeout period."
    },
    "InvalidFormatException": {
        "description": "Related to JSON deserialization, occurs when data cannot be deserialized into the expected format."
    },
    "RuntimeException": {
        "description": "A general runtime exception, indicating an unexpected issue during execution."
    },
    "ClassNotFoundException": {
        "description": "Thrown when a class is not found, suggesting a missing or incorrectly named class."
    },
    "NullPointerException": {
        "description": "Indicates a null pointer exception, often due to accessing a member of a null object."
    },
    "Other": {
        "description": "Any other error not covered by the existing categories."
    }
}


describe.each(config.languages.map(o => [o.name, o] as [string, Language]))("Testing language %s", (name, lang) => {
    const cwd = lang.snippetRoot(workspaceDir)
    const filePaths = glob.globSync(lang.glob, { cwd })

    test.concurrent.each(filePaths)('testing file %s', async (filePath) => {
        const { error } = await execCmd(lang.exec(filePath), cwd)

        const co = new CohereClient()

        let err = { "error": "passed" }

        if (error) {
            const explain = await co.v2.chat({
                model: "command-r-plus",
                messages: [{
                    role: "user", content: `
                    Please classify this error into the following categories:
                    ${error}
                    
                    error categories:
    
                    ${JSON.stringify(failureReasons)}`
                }],
                temperature: 0,
                responseFormat: {
                    type: "json_object",
                    jsonSchema: {
                        type: "object",
                        required: ["error"],
                        properties: {
                            error: {
                                type: "string",
                                enum: Object.keys(failureReasons)
                            }
                        }
                    }
                }
            })

            err = JSON.parse(explain.message?.content?.[0].text || "{}")
            console.error(error)
        }
        
        expect(err.error).toMatchSnapshot()

    }, 1000000);
})