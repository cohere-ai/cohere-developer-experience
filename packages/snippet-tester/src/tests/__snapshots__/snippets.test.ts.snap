// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing language golang testing file chat-post/default/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file chat-post/documents/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file chat-post/stream/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file chat-post/tools/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file classify-post/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file connector-create/main.go 1`] = `"BadRequestError"`;

exports[`Testing language golang testing file connector-delete/main.go 1`] = `"InternalServerError"`;

exports[`Testing language golang testing file connector-get/main.go 1`] = `"BadRequestError"`;

exports[`Testing language golang testing file connector-patch/main.go 1`] = `"InternalServerError"`;

exports[`Testing language golang testing file connectors-id-oauth-authorize-post/main.go 1`] = `"NotFoundError"`;

exports[`Testing language golang testing file connectors-list/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file dataset-delete/main.go 1`] = `"BadRequestError"`;

exports[`Testing language golang testing file dataset-get/main.go 1`] = `"NotFoundError"`;

exports[`Testing language golang testing file dataset-post/main.go 1`] = `"BadRequestError"`;

exports[`Testing language golang testing file dataset-usage-get/main.go 1`] = `"JsonError"`;

exports[`Testing language golang testing file detokenize-post/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file embed-jobs-cancel/main.go 1`] = `"NotFoundError"`;

exports[`Testing language golang testing file embed-jobs-get/main.go 1`] = `"NotFoundError"`;

exports[`Testing language golang testing file embed-jobs-post/main.go 1`] = `"BadRequestError"`;

exports[`Testing language golang testing file embed-post/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file emebed-jobs-get/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file finetuning/create-finetuned-model/main.go 1`] = `"BadRequestError"`;

exports[`Testing language golang testing file finetuning/delete-finetuned-model/main.go 1`] = `"BadRequestError"`;

exports[`Testing language golang testing file finetuning/get-finetuned-model/main.go 1`] = `undefined`;

exports[`Testing language golang testing file finetuning/list-events/main.go 1`] = `"InternalServerError"`;

exports[`Testing language golang testing file finetuning/list-finetuned-models/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file finetuning/list-training-step-metrics/main.go 1`] = `"BadRequestError"`;

exports[`Testing language golang testing file finetuning/update-finetuned-model/main.go 1`] = `"BadRequestError"`;

exports[`Testing language golang testing file generate-post/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file models-list-get/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file rerank-post/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file summarize-post/main.go 1`] = `"passed"`;

exports[`Testing language golang testing file tokenize-post/main.go 1`] = `"passed"`;

exports[`Testing language java testing file app/src/main/java/ClassifyPost.java 1`] = `"ClassNotFoundException"`;

exports[`Testing language java testing file app/src/main/java/ConnectorCreate.java 1`] = `"BadRequestError"`;

exports[`Testing language java testing file app/src/main/java/ConnectorDelete.java 1`] = `"NotFoundError"`;

exports[`Testing language java testing file app/src/main/java/ConnectorGet.java 1`] = `"NotFoundError"`;

exports[`Testing language java testing file app/src/main/java/ConnectorPatch.java 1`] = `"NotFoundError"`;

exports[`Testing language java testing file app/src/main/java/ConnectorsIdOauthAuthorizePost.java 1`] = `"NotFoundError"`;

exports[`Testing language java testing file app/src/main/java/ConnectorsList.java 1`] = `"passed"`;

exports[`Testing language java testing file app/src/main/java/DatasetDelete.java 1`] = `"NotFoundError"`;

exports[`Testing language java testing file app/src/main/java/DatasetGet.java 1`] = `"NotFoundError"`;

exports[`Testing language java testing file app/src/main/java/DatasetPost.java 1`] = `"RuntimeException"`;

exports[`Testing language java testing file app/src/main/java/DatasetUsageGet.java 1`] = `"InternalServerError"`;

exports[`Testing language java testing file app/src/main/java/DetokenizePost.java 1`] = `"passed"`;

exports[`Testing language java testing file app/src/main/java/EmbedJobsCancel.java 1`] = `"NotFoundError"`;

exports[`Testing language java testing file app/src/main/java/EmbedJobsGet.java 1`] = `"passed"`;

exports[`Testing language java testing file app/src/main/java/EmbedJobsPost.java 1`] = `"NotFoundError"`;

exports[`Testing language java testing file app/src/main/java/EmbedPost.java 1`] = `"passed"`;

exports[`Testing language java testing file app/src/main/java/EmebedJobsGet.java 1`] = `"NotFoundError"`;

exports[`Testing language java testing file app/src/main/java/GeneratePost.java 1`] = `"InternalServerError"`;

exports[`Testing language java testing file app/src/main/java/ModelsListGet.java 1`] = `"InvalidFormatException"`;

exports[`Testing language java testing file app/src/main/java/RerankPost.java 1`] = `"passed"`;

exports[`Testing language java testing file app/src/main/java/SummarizePost.java 1`] = `"passed"`;

exports[`Testing language java testing file app/src/main/java/TokenizePost.java 1`] = `"passed"`;

exports[`Testing language java testing file app/src/main/java/chatpost/Default.java 1`] = `"passed"`;

exports[`Testing language java testing file app/src/main/java/chatpost/Documents.java 1`] = `"passed"`;

exports[`Testing language java testing file app/src/main/java/chatpost/Stream.java 1`] = `"InternalServerError"`;

exports[`Testing language java testing file app/src/main/java/chatpost/Tools.java 1`] = `"passed"`;

exports[`Testing language java testing file app/src/main/java/finetuning/CreateFinetunedModel.java 1`] = `"BadRequestError"`;

exports[`Testing language java testing file app/src/main/java/finetuning/DeleteFinetunedModel.java 1`] = `"BadRequestError"`;

exports[`Testing language java testing file app/src/main/java/finetuning/GetFinetunedModel.java 1`] = `"BadRequestError"`;

exports[`Testing language java testing file app/src/main/java/finetuning/ListEvents.java 1`] = `"ClassNotFoundException"`;

exports[`Testing language java testing file app/src/main/java/finetuning/ListFinetunedModels.java 1`] = `"passed"`;

exports[`Testing language java testing file app/src/main/java/finetuning/ListTrainingStepMetrics.java 1`] = `"BadRequestError"`;

exports[`Testing language java testing file app/src/main/java/finetuning/UpdateFinetunedModel.java 1`] = `"BadRequestError"`;

exports[`Testing language node testing file chat-post/default.js 1`] = `"passed"`;

exports[`Testing language node testing file chat-post/documents.js 1`] = `"passed"`;

exports[`Testing language node testing file chat-post/stream.js 1`] = `"TypeError"`;

exports[`Testing language node testing file chat-post/tools.js 1`] = `"passed"`;

exports[`Testing language node testing file classify-post.js 1`] = `"passed"`;

exports[`Testing language node testing file connector-create.js 1`] = `"BadRequestError"`;

exports[`Testing language node testing file connector-delete.js 1`] = `"NotFoundError"`;

exports[`Testing language node testing file connector-get.js 1`] = `"NotFoundError"`;

exports[`Testing language node testing file connector-patch.js 1`] = `"ReferenceError"`;

exports[`Testing language node testing file connectors-id-oauth-authorize-post.js 1`] = `"NotFoundError"`;

exports[`Testing language node testing file connectors-list.js 1`] = `"passed"`;

exports[`Testing language node testing file dataset-get.js 1`] = `"NotFoundError"`;

exports[`Testing language node testing file dataset-post.js 1`] = `"ReferenceError"`;

exports[`Testing language node testing file dataset-usage-get.js 1`] = `"passed"`;

exports[`Testing language node testing file detokenize-post.js 1`] = `"passed"`;

exports[`Testing language node testing file embed-jobs-cancel.js 1`] = `"NotFoundError"`;

exports[`Testing language node testing file embed-jobs-get.js 1`] = `"passed"`;

exports[`Testing language node testing file embed-jobs-post.js 1`] = `"NotFoundError"`;

exports[`Testing language node testing file embed-post.js 1`] = `"passed"`;

exports[`Testing language node testing file emebed-jobs-get.js 1`] = `"NotFoundError"`;

exports[`Testing language node testing file finetuning/create-finetuned-model.js 1`] = `"TypeError"`;

exports[`Testing language node testing file finetuning/delete-finetuned-model.js 1`] = `"BadRequestError"`;

exports[`Testing language node testing file finetuning/get-finetuned-model.js 1`] = `"BadRequestError"`;

exports[`Testing language node testing file finetuning/list-events.js 1`] = `"InternalServerError"`;

exports[`Testing language node testing file finetuning/list-finetuned-models.js 1`] = `"passed"`;

exports[`Testing language node testing file finetuning/list-training-step-metrics.js 1`] = `"BadRequestError"`;

exports[`Testing language node testing file finetuning/update-finetuned-model.js 1`] = `"JsonError"`;

exports[`Testing language node testing file generate-post.js 1`] = `"passed"`;

exports[`Testing language node testing file models-list-get.js 1`] = `"passed"`;

exports[`Testing language node testing file rerank-post.js 1`] = `"passed"`;

exports[`Testing language node testing file summarize-post.js 1`] = `"passed"`;

exports[`Testing language node testing file tokenize-post.js 1`] = `"passed"`;

exports[`Testing language python testing file chat-post.py 1`] = `"passed"`;

exports[`Testing language python testing file chat-post/default.py 1`] = `"passed"`;

exports[`Testing language python testing file chat-post/documents.py 1`] = `"passed"`;

exports[`Testing language python testing file chat-post/stream.py 1`] = `"passed"`;

exports[`Testing language python testing file chat-post/tools.py 1`] = `"passed"`;

exports[`Testing language python testing file classify-post.py 1`] = `"passed"`;

exports[`Testing language python testing file connector-create.py 1`] = `"BadRequestError"`;

exports[`Testing language python testing file connector-delete.py 1`] = `"NotFoundError"`;

exports[`Testing language python testing file connector-get.py 1`] = `"NotFoundError"`;

exports[`Testing language python testing file connector-patch.py 1`] = `"TypeError"`;

exports[`Testing language python testing file connectors-id-oauth-authorize-post.py 1`] = `"TypeError"`;

exports[`Testing language python testing file connectors-list.py 1`] = `"passed"`;

exports[`Testing language python testing file dataset-delete.py 1`] = `"NotFoundError"`;

exports[`Testing language python testing file dataset-get.py 1`] = `"NotFoundError"`;

exports[`Testing language python testing file dataset-post.py 1`] = `"FileNotFoundError"`;

exports[`Testing language python testing file dataset-usage-get.py 1`] = `"passed"`;

exports[`Testing language python testing file detokenize-post.py 1`] = `"passed"`;

exports[`Testing language python testing file embed-jobs-cancel.py 1`] = `"NotFoundError"`;

exports[`Testing language python testing file embed-jobs-get.py 1`] = `"passed"`;

exports[`Testing language python testing file embed-jobs-post.py 1`] = `"NotFoundError"`;

exports[`Testing language python testing file embed-post.py 1`] = `"passed"`;

exports[`Testing language python testing file emebed-jobs-get.py 1`] = `"NotFoundError"`;

exports[`Testing language python testing file finetuning/create-finetuned-model.py 1`] = `"BadRequestError"`;

exports[`Testing language python testing file finetuning/delete-finetuned-model.py 1`] = `"BadRequestError"`;

exports[`Testing language python testing file finetuning/get-finetuned-model.py 1`] = `"BadRequestError"`;

exports[`Testing language python testing file finetuning/list-events.py 1`] = `"InternalServerError"`;

exports[`Testing language python testing file finetuning/list-finetuned-models.py 1`] = `"passed"`;

exports[`Testing language python testing file finetuning/list-training-step-metrics.py 1`] = `"BadRequestError"`;

exports[`Testing language python testing file finetuning/update-finetuned-model.py 1`] = `"BadRequestError"`;

exports[`Testing language python testing file generate-post.py 1`] = `"passed"`;

exports[`Testing language python testing file models-list-get.py 1`] = `"passed"`;

exports[`Testing language python testing file rerank-post.py 1`] = `"passed"`;

exports[`Testing language python testing file summarize-post.py 1`] = `"passed"`;

exports[`Testing language python testing file tokenize-post.py 1`] = `"passed"`;
